pkg/api/handlers/projectinvitation.go

```diff
@@ -0,0 +1,332 @@
+package handlers
+
+import (
+	"strings"
+	"time"
+
+	"github.com/google/uuid"
+	"github.com/obot-platform/obot/apiclient/types"
+	"github.com/obot-platform/obot/pkg/api"
+	v1 "github.com/obot-platform/obot/pkg/storage/apis/obot.obot.ai/v1"
+	"github.com/obot-platform/obot/pkg/system"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type ProjectInvitationHandler struct{}
+
+func NewProjectInvitationHandler() *ProjectInvitationHandler {
+	return &ProjectInvitationHandler{}
+}
+
+func (h *ProjectInvitationHandler) CreateInvitationForProject(req api.Context) error {
+	thread, err := getProjectThread(req)
+	if err != nil {
+		return err
+	}
+
+	if !thread.Spec.Project {
+		return types.NewErrBadRequest("only projects can have invitations")
+	}
+
+	if thread.Spec.UserID != req.User.GetUID() {
+		return types.NewErrForbidden("only the project creator can create invitations")
+	}
+
+	var project types.Project
+	if thread.Spec.ParentThreadName != "" {
+		var parentThread *v1.Thread
+		if err := req.Get(parentThread, thread.Spec.ParentThreadName); err != nil {
+			return err
+		}
+		project = convertProject(thread, parentThread)
+	} else {
+		project = convertProject(thread, nil)
+	}
+
+	// Generate a random code for the invitation
+	code := strings.ReplaceAll(uuid.New().String(), "-", "")
+
+	invitation := v1.ProjectInvitation{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      code,
+			Namespace: req.Namespace(),
+		},
+		Spec: v1.ProjectInvitationSpec{
+			Status:   string(types.ProjectInvitationStatusPending),
+			ThreadID: thread.Name,
+		},
+	}
+
+	if err := req.Create(&invitation); err != nil {
+		return err
+	}
+
+	return req.WriteCreated(types.ProjectInvitationManifest{
+		Code:    code,
+		Project: &project,
+		Status:  types.ProjectInvitationStatusPending,
+		Created: invitation.CreationTimestamp.Format(time.RFC3339),
+	})
+}
+
+func (h *ProjectInvitationHandler) ListInvitationsForProject(req api.Context) error {
+	thread, err := getProjectThread(req)
+	if err != nil {
+		return err
+	}
+
+	if !thread.Spec.Project {
+		return types.NewErrBadRequest("only projects can have invitations")
+	}
+
+	if thread.Spec.UserID != req.User.GetUID() {
```

Comment:
Same question here about admins doing this.

Reply:
Yes, fixed